### PR TITLE
Rocks db

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
The RocksDb was reverted

#### Summary of Changes

Reintroduce Rocksdb module and changes. New changes since last time are introduced in the latest commit.

Fixes #

### CODE REVIEW

**Comment by carllin** [comment]
Oops, forgot to run it this time, sorry!

File: src/db_ledger.rs [hunk 1]
```diff
@@ -0,0 +1,622 @@
...
+            db,
+            meta_cf,
+            data_cf,
+            erasure_cf,
+        })
+    }
+
+    pub fn write_blobs<'a, I>(&mut self, slot: u64, blobs: &'a I) -> Result<()>
+    where
+        &'a I: IntoIterator<Item = &'a &'a Blob>,
```
Comment (rob-solana) [inline src/db_ledger.rs]: this reads like something I've seen in ledger, block, etc.  I don't fully understand the constraints on I, but I assume it's as flexible as possible (takes vecs, slices, iterators, etc.)

File: src/db_ledger.rs [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+            db_iterator.next();
+        }
+
+        Ok((total_blobs, total_current_size as u64))
+    }
+
+    fn is_blob_data_valid(blob: &Blob) -> bool {
+        if let Ok(blob_size) = blob.size() {
+            let blob_data = &blob.data[BLOB_HEADER_SIZE..BLOB_HEADER_SIZE + blob_size];
+            let entry: BincodeResult<Entry> = deserialize(blob_data);
```
Comment (rob-solana) [inline src/db_ledger.rs]: misnamed, I think.  you're asking whether the blob can be an entry, which is possibly beyond the purview of this module...

you might add a trait to what gets stored or just bail on this?  what are you defending against with this check?

File: src/db_ledger.rs (hunk 1 above)
Comment (garious) [reply src/db_ledger.rs]: The double ref is surprising

Comment (carllin) [reply src/db_ledger.rs]: This was engineered to accept slices xD. So you could pass it references to vectors which coerce down to the appropriate type, or pass a slice. The double ref is due to how slices implement IntoIterator: https://stackoverflow.com/questions/35144386/passing-slice-as-an-intoiterator

Comment (garious) [reply src/db_ledger.rs]: Amazing, thanks!

**Pull request merged**